### PR TITLE
Upgrade to Phoenix 5.3.1.0

### LINKED ISSUE #54: Update to 5.3.1.0
Working on it.

### PR DESCRIPTION
**Pull request by virtuald**
- Fixes #54

@auscompgeek @ariovistus can you test this on real hardware and your sims? Wheel can be found at https://www.tortall.net/~robotpy/wheels/2018/robotpy_ctre-2018.2.5.post0.dev1-cp36-cp36m-linux_armv7l.whl

### CODE REVIEW

File: ctre/basemotorcontroller.py [hunk 1]
```diff
@@ -93,18 +96,40 @@ def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0):
...
+            * Neutral: Ignore demand1 and apply no change to the demand0 output.
+            * AuxPID: Use demand1 to set the target for the auxiliary PID 1.
+            * ArbitraryFeedForward: Use demand1 as an arbitrary additive value to the
+              demand0 output.  In PercentOutput the demand0 output is the motor output,
+              and in closed-loop modes the demand0 output is the output of PID0.
+    
+        Arcade Drive Example::
+        
+            _talonLeft.set(ControlMode.PercentOutput, joyForward, DemandType.ArbitraryFeedForward, +joyTurn)
+            _talonRght.set(ControlMode.PercentOutput, joyForward, DemandType.ArbitraryFeedForward, -joyTurn)
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: I think the arguments here are the wrong way around. Or maybe we should have it this way around in the signature — I don't think it'd make sense to not specify the type of demand1 if you're passing it in anyway.

File: ctre/basemotorcontroller.py [hunk 2]
```diff
@@ -72,14 +75,14 @@ def getDeviceID(self):
         """
         return self.getDeviceNumber()
 
-    def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0):
+    def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0, demand1Type: DemandType = DemandType.Neutral):
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: Now that this takes 4 arguments, our WPI wrapper classes need to acknowledge this. I personally think using `match_arglist` there is mostly a bad idea anyway.

File: ctre/basemotorcontroller.py [hunk 3]
```diff
@@ -231,7 +264,15 @@ def follow(self, masterToFollow: 'BaseMotorController'):
         id24 = id24 & 0xFFFF
         id24 <<= 8
         id24 |= (id32 & 0xFF)
-        self.set(ControlMode.Follower, id24)
+        
+        if followerType == FollowerType.PercentOutput:
+            self.set(ControlMode.Follower, id24)
+        elif followerType == FollowerType.AuxOutput1:
+            # follow the motor controller, but set the aux flag
+		    # to ensure we follow the processed output
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: Heh, I see tabs here.

File: ctre/basemotorcontroller.py [hunk 4]
```diff
@@ -114,9 +139,9 @@ def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0):
                 work |= int(demand0) & 0xFF
             else:
                 work = int(demand0)
-            self.setDemand(self.sendMode, work, 0)
-        elif self.controlMode in [ControlMode.Velocity, ControlMode.Position, ControlMode.MotionMagic, ControlMode.MotionMagicArc, ControlMode.MotionProfile]:
-            self.setDemand(self.sendMode, int(demand0), 0)
+            self._set_4(self.sendMode, demand0, demand1, demand1Type)
+        elif self.controlMode in [ControlMode.Velocity, ControlMode.Position, ControlMode.MotionMagic, ControlMode.MotionProfile, ControlMode.MotionProfileArc]:
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: I'm thinking this should probably be a set. I guess it doesn't really matter too much though.

File: ctre/basemotorcontroller.py [hunk 5]
```diff
@@ -93,18 +96,40 @@ def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0):
...
+        
+            # Note: Other configurations (sensor selection, PID gains, etc.) need to be set.
+            _talonLeft.follow(_talonRght, FollowerType.AuxOutput1);
+            _talonRght.set(ControlMode.MotionMagic, targetDistance, DemandType.AuxPID, desiredRobotHeading)
+        
         """
         self.controlMode = mode
         self.sendMode = mode
 
         if self.controlMode == ControlMode.PercentOutput:
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: Hm. I just realised, shouldn't these comparisons be identity checks (`is`) instead?

File: gen/MotController_data.yml [hunk 6]
```diff
@@ -794,6 +827,41 @@ c_MotController_ConfigMaxIntegralAccumulator:
...
+                       the magnitude will apply in both forward and reverse directions.
+    
+    :param timeoutMs:  Timeout value in ms. If nonzero, function will wait for
+                       config success and report an error if it times out.
+                       If zero, no blocking or checking is performed.
+    
+    :returns: Error Code generated by function. 0 indicates no error.
+c_MotController_ConfigClosedLoopPeriod:
+  code: |
+    self.hal_data['profile%d_period' % slotIdx] = loopTimeMs / 1000.0
```
Comment (auscompgeek) [inline gen/MotController_data.yml]: I guess this should be used by the simulation PID loop, probably.

**Comment by ariovistus** [comment]
doesn't seem to break my sims

File: gen/MotController_data.yml (hunk 6 above)
Comment (ariovistus) [reply gen/MotController_data.yml]: aw, man did they make that configurable?

File: ctre/basemotorcontroller.py (hunk 5 above)
Comment (virtuald) [reply ctre/basemotorcontroller.py]: I think `==` is fine.

File: ctre/basemotorcontroller.py [hunk 7]
```diff
@@ -71,15 +75,17 @@ def getDeviceID(self):
         :returns: Device number.
         """
         return self.getDeviceNumber()
+        
+    __set4_modes = set([ControlMode.Velocity, ControlMode.Position, ControlMode.MotionMagic, ControlMode.MotionProfile, ControlMode.MotionProfileArc])
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: Why not use the set literal syntax here?

File: ctre/basemotorcontroller.py [hunk 8]
```diff
@@ -114,9 +142,9 @@ def set(self, mode: ControlMode, demand0: float, demand1: float = 0.0):
                 work |= int(demand0) & 0xFF
             else:
                 work = int(demand0)
-            self.setDemand(self.sendMode, work, 0)
-        elif self.controlMode in [ControlMode.Velocity, ControlMode.Position, ControlMode.MotionMagic, ControlMode.MotionMagicArc, ControlMode.MotionProfile]:
-            self.setDemand(self.sendMode, int(demand0), 0)
+            self._set_4(self.sendMode, demand0, demand1, demand1Type)
```
Comment (auscompgeek) [inline ctre/basemotorcontroller.py]: This should be `work` instead of `demand0`.

File: ctre/basemotorcontroller.py (hunk 5 above)
Comment (auscompgeek) [reply ctre/basemotorcontroller.py]: Well, realistically we don't actually support passing ints here, so calling `__eq__` seems wasteful.

File: ctre/basemotorcontroller.py (hunk 8 above)
Comment (virtuald) [reply ctre/basemotorcontroller.py]: Good catch.

**Pull request merged**